### PR TITLE
fix(shareReplay): handle possible memory leaks

### PR DESCRIPTION
**Pull request by josepot**
**Description:**
Fixes possible memory leaks caused by the `shareReplay` operator

**Related issue (if exists):**
#5931

### CODE REVIEW

**Comment by josepot** [comment]
I tested this locally, and the changes in this PR don't fix the memory leak reported on #5931, so I'm closing it for now. I wil re-open it if I find a solution.

**Comment by josepot** [comment]
Actually, sorry, [it *does* fix the memory leak](https://stackblitz.com/edit/angular-7jmpmo?file=src%2Fapp%2Fvalue.service.ts)

**Comment by MaximSagan** [comment]
I believe this PR should be merged as there is no reason to keep those subscription objects around any longer than necessary, but I think it doesn't get to the root of the problem here.

That is, we have to ask why an instance of `Subscriber` (which is what `Observable.prototype.subscribe(observer)` returns) is maintaining any reference at all to its `Observer` after unsubscription (including complete & error).

With that in mind, I have made this related PR, so that this kind of thing will (hopefully) never be a problem again: #5934

File: src/internal/operators/shareReplay.ts [hunk 1]
```diff
@@ -108,13 +108,17 @@ function shareReplayOperator<T>({
...
+        subscription = undefined;
+      }
     } else {
       innerSub = subject.subscribe(this);
     }
 
     this.add(() => {
       refCount--;
       innerSub.unsubscribe();
+      innerSub = undefined;
```
Comment (benlesh) [inline src/internal/operators/shareReplay.ts:121]: This isn't necessary, really. As there would be no more references to `innerSub` held after this, and it would get GC'ed.

File: src/internal/operators/shareReplay.ts [hunk 2]
```diff
@@ -108,13 +108,17 @@ function shareReplayOperator<T>({
           subject.complete();
         },
       });
+      if (isComplete) {
+        subscription = undefined;
+      }
```
Comment (benlesh) [inline src/internal/operators/shareReplay.ts:113]: I'm not sure how this solves a memory leak. Do we have tests or proof for this?

The *only* way this conditional passes is if the `source` completed synchronously. Which, if you [look right here](https://github.com/ReactiveX/rxjs/blob/1e1e022ed9fdc4a4bfed63cf151887031ac57398/src/internal/operators/shareReplay.ts#L106-L107), the `isComplete` is set to `true` and the `subscription` is set to `undefined` during that very same completion. 

In other words, this work has already been done by the time that `isComplete` would ever be `true`. So we can remove this statement entirely.

**Review by benlesh — COMMENTED** [review]
These changes don't at all affect the semantics of what's going on in this operator, and wouldn't really effect the memory usage. 

As I stated in the related issue, it seems the problem is confusion around the default behavior of `shareReplay`. Which, IMO, has been an ongoing problem.

Hopefully the new operators in version 7 will gain more adoption and we can leave these confusing days behind us.

**Comment by benlesh** [comment]
I'm going to close this PR because the changes don't really do anything (as I outlined above). If there's proof that this changes the memory usage, I'll be happy to take another look.

File: src/internal/operators/shareReplay.ts (hunk 2 above)
Comment (josepot) [reply src/internal/operators/shareReplay.ts:113]: > I'm not sure how this solves a memory leak. Do we have tests or proof for this?

Yes, we do, in [this link](https://stackblitz.com/edit/angular-7jmpmo?file=src%2Fapp%2Fvalue.service.ts). If you follow the reproduction steps of [this comment](https://github.com/ReactiveX/rxjs/issues/5931#issue-765399108), then you will notice that the memory leak is not present with these changes.

> The only way this conditional passes is if the source completed synchronously. 

Exactly!

> Which, if you look right here, the isComplete is set to true and the subscription is set to undefined during that very same completion.

That's what it _looks_ like. However, that's not what it actually happens... The thing is that when it completes synchronously, then in that moment `subscription` is `undefined`, because its value has not been assigned yet. It will be assigned after the callback gets executed. Therefore, the `if` statement is necessary to prevent the memory leak.

**Comment by josepot** [comment]
I'm sorry if I'm being annoying, but @benlesh or @cartant could you please have a look at [this codesandbox](https://codesandbox.io/s/beautiful-wood-m2h9o?file=/src/index.ts) :pray:. It's simply doing a `of(1).pipe(shareReplay()).subscribe()` while using the current version of `shareReplay` on v6 with 2 `console.log`s that illustrate why the `if (isComplete)` statement that I suggested in this PR is necessary in order to avoid the memory leak.

Also, thanks a lot for the amazing work that you are doing on v7. I've been using it on a project of mine and so far things are looking great! There is just one little thing that's annoying me: which is that I'm seeing code in a chunk that I really think that it should have been "tree shaken" and it wasn't. I'm still investigating, I will open an issue or a discussion in the coming days to explain this better and I will make sure that when I open the issue I can give you as much info as I can. Thanks again!

**Comment by cartant** [comment]
@josepot I'll have a look at this in a bit. I'll push a test to this branch to clear things up.

**Comment by benlesh** [comment]
Derp. @josepot is correct. I overlooked that synchronously, `subscription` is set *after* it's set to `undefined` when the `subscribe` call returns. Reopening this.

File: src/internal/operators/shareReplay.ts (hunk 2 above)
Comment (benlesh) [reply src/internal/operators/shareReplay.ts:113]: Yeah. I was wrong. Sorry about that, I should have looked more carefully. Thank you for being diligent.

Here's what would have avoided my mistake: A detailed comment above this check like:

```ts
// Here we need to check to see if the source synchronously completed. Although
// we're setting `subscription = undefined` in the completion handler, if the source
// is synchronous, that will happen *before* subscription is set by the return of
// the `subscribe` call.
```

**Review by benlesh — CHANGES_REQUESTED** [review]
1. Add a comment, as suggested, above the new conditional.

2. Add a test that proves this resolves the issue.

RELATED: we'll need to create a second PR that adds the same test to `master`.

File: src/internal/operators/shareReplay.ts (hunk 2 above)
Comment (josepot) [reply src/internal/operators/shareReplay.ts:113]: No worries! I should have added a comment in the first place, since this is very counter intuitive.

**Comment by josepot** [comment]
> Add a comment, as suggested, above the new conditional.

:heavy_check_mark: 

> Add a test that proves this resolves the issue.

Is that possible without testing the internals? :thinking: I mean, the only way that I can think off, involves exposing an implementation detail... which IMO wouldn't be worth it. I'm more than open to suggestions.

**Comment by cartant** [comment]
> Is that possible without testing the internals? 🤔

Yes. I'll add one soon.

**Comment by cartant** [comment]
@josepot @benlesh I've added a leak test for `shareReplay` to effect a failure, you will have to comment out the nulling of `innerSub` and `subscription` (although, the latter is, IMO, a smaller leak that is not related to the problem demonstrated in the Angular StackBlitz repro). Note that `FinalizationRegistry` is only in Node 14 and later. I'll add a similar test to `master` soon.

**Comment by cartant** [comment]
IDK what these build failures are. I hate CI so much.

**Comment by josepot** [comment]
@cartant hats off to you for that test!! :clap: :clap: :clap: TIL about [`FinalizationRegistry`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry), amazing! Thanks a lot!

PS @voliva [check this out](https://github.com/ReactiveX/rxjs/pull/5932/files#diff-3b05f28653d60d8c6acca2c70bee58c9c8dad00078f05a7e363e6f9253dd792d) we should start using this on `@react-rxjs/core` ASAP.

**Comment by cartant** [comment]
@josepot IDK whether or not you can see the details of the CI failure, but this is what's breaking:

```
Error: node_modules/@types/node/index.d.ts(74,11): error TS2300: Duplicate identifier 'IteratorResult'.
Error: node_modules/typescript/lib/lib.es2015.iterable.d.ts(41,6): error TS2300: Duplicate identifier 'IteratorResult'.
npm ERR! code 2
npm ERR! path /home/runner/work/rxjs/rxjs
npm ERR! command failed
npm ERR! command sh -c tsc -p ./tsconfig/compat/tsconfig.cjs.json
```

🤷‍♂️ Maybe there is was a breaking change in the node types that match whatever semver is in this branch's `package.json`. IDK.

**Comment by josepot** [comment]
> @josepot IDK whether or not you can see the details of the CI failure, but this is what's breaking:
> 
> ```
> Error: node_modules/@types/node/index.d.ts(74,11): error TS2300: Duplicate identifier 'IteratorResult'.
> Error: node_modules/typescript/lib/lib.es2015.iterable.d.ts(41,6): error TS2300: Duplicate identifier 'IteratorResult'.
> npm ERR! code 2
> npm ERR! path /home/runner/work/rxjs/rxjs
> npm ERR! command failed
> npm ERR! command sh -c tsc -p ./tsconfig/compat/tsconfig.cjs.json
> ```
> 
>  Maybe there is was a breaking change in the node types that match whatever semver is in this branch's `package.json`. IDK.

Thanks a lot @cartant ! Yep, I can see that. There were 2 different checks failing, one has been fixed by my latest commit (a linting issue, aparently). And this one, I'm investigating... The first thing that I tried was rebasing the branch from 6.x, just in case, but that didn't help... And yeah, based on a quick google search that error normally indicates an issue with the node types. I will investigate a bit more later. Thanks again.

**Comment by josepot** [comment]
> man_shrugging Maybe there is was a breaking change in the node types that match whatever semver is in this branch's package.json. IDK.

Yep, I think that's the issue... Because -according to the logs- the task that fails is running under node@13.14.0:

```sh
/usr/bin/tar xz --warning=no-unknown-keyword -C /home/runner/work/_temp/7bd7433d-ba81-4237-ac22-1ea42ffa579b -f /home/runner/work/_temp/1e3c4cf6-32b1-405a-90d3-3bb7230d395f
/opt/hostedtoolcache/node/13.14.0/x64/bin/node --version
v13.14.0
```

And the types that are defined on the `package-lock.json` file are `9.4.5`.

Question: would it make sense to add the `skipLibCheck` for this task  `tsc -p ./tsconfig/compat/tsconfig.cjs.json`?

**Review by cartant — APPROVED** [review]

File: spec/operators/shareReplay-spec.ts [hunk 3]
```diff
@@ -262,4 +262,26 @@ describe('shareReplay operator', () => {
     expectObservable(result).toBe(expected);
   });
 
+  const FinalizationRegistry = (global as any).FinalizationRegistry;
```
Comment (benlesh) [inline spec/operators/shareReplay-spec.ts:265]: This is a clever use of `FinalizationRegistry`.

**Review by benlesh — APPROVED** [review]

**Comment by benlesh** [comment]
Great work, @josepot and @cartant. Thanks for seeing this through!
